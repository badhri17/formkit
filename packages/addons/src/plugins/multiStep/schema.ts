import { FormKitTypeDefinition } from '@formkit/core'
import { $if, defaultIcon, localize, disablesChildren } from '@formkit/inputs'
import {
  wrapper,
  badge,
  stepPrevious,
  stepNext,
  stepOuter,
  tab,
  tabLabel,
  tabs,
  steps,
  stepActions,
  stepInner,
  stepIcon,
  multiStepOuter,
} from './sections'

export const multiStep: FormKitTypeDefinition = {
  /**
   * The actual schema of the input, or a function that returns the schema.
   */
  schema: multiStepOuter(
    wrapper(
      tabs(
        tab(
          $if(
            '$tabStyle === "tab" || ($tabStyle === "progress" && $hideProgressLabels === false)',
            tabLabel('$step.stepName')
          ),
          $if(
            '($step.totalErrorCount > 0) && $step.showStepErrors',
            badge('$step.totalErrorCount')
          ),
          $if(
            '$step.isValid && $step.hasBeenVisited',
            badge(stepIcon('validStep'))
          )
        )
      ),
      steps('$slots.default')
    )
  ),
  /**
   * The type of node, can be a list, group, or input.
   */
  type: 'group',
  /**
   * The family of inputs this one belongs too. For example "text" and "email"
   * are both part of the "text" family. This is primary used for styling.
   */
  family: 'multi-step',
  /**
   * An array of extra props to accept for this input.
   */
  props: [
    'allowIncomplete',
    'hideProgressLabels',
    'tabStyle',
    'beforeStepChange',
    'validStepIcon',
  ],
  /**
   * Additional features that should be added to your input
   */
  features: [defaultIcon('validStep', 'check'), disablesChildren],
}

export const step: FormKitTypeDefinition = {
  /**
   * The actual schema of the input, or a function that returns the schema.
   */
  schema: stepOuter(
    stepInner('$slots.default'),
    stepActions(stepPrevious(), stepNext())
  ),
  /**
   * The type of node, can be a list, group, or input.
   */
  type: 'group',
  /**
   * The family of inputs this one belongs too. For example "text" and "email"
   * are both part of the "text" family. This is primary used for styling.
   */
  family: '',
  /**
   * An array of extra props to accept for this input.
   */
  props: [
    'previousLabel',
    'nextLabel',
    'beforeStepChange',
    'previousAttrs',
    'nextAttrs',
    'validStepIcon',
  ],
  /**
   * Additional features that should be added to your input
   */
  features: [localize('next'), localize('prev'), disablesChildren],
}
