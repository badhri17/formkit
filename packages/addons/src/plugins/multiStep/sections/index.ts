export { wrapper } from './wrapper'
export { badge } from './badge'
export { stepActions } from './stepActions'
export { stepInner } from './stepInner'
export { stepNext } from './stepNext'
export { stepOuter } from './stepOuter'
export { stepPrevious } from './stepPrevious'
export { steps } from './steps'
export { stepIcon } from './stepIcon'
export { tab } from './tab'
export { tabLabel } from './tabLabel'
export { tabs } from './tabs'
export { multiStepOuter } from './multiStepOuter'
