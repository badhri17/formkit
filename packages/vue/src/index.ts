import type { FormKitRootConfig } from '@formkit/core'
/**
 * The official FormKit/Vue integration. This package is responsible for
 * integrating Vue with FormKit core and other first-party packages.
 *
 *
 * @packageDocumentation
 */

declare global {
  // eslint-disable-next-line no-var
  var __FORMKIT_CONFIGS__: FormKitRootConfig[]
}

/**
 * The useInput composable.
 */
export { useInput } from './composables/useInput'

/**
 * Shorthand for creating inputs with standard FormKit features.
 */
export { createInput } from './composables/createInput'

/**
 * The defineFormKitConfig composable.
 */
export { defineFormKitConfig } from './composables/defineFormKitConfig'

/**
 * The plugin and plugin types.
 */
export * from './plugin'

/**
 * The root FormKit component.
 */
export {
  default as FormKit,
  FormKitComponent,
  FormKitSetupContext,
  Slots,
  parentSymbol,
  componentSymbol,
  getCurrentSchemaNode,
} from './FormKit'

/**
 * The FormKitRoot wrapper component used to provide context to FormKit
 * about whether a FormKit input is booting in a Document or ShadowRoot.
 */
export { FormKitRoot, rootSymbol } from './FormKitRoot'

/**
 * The FormKitKitchenSink component.
 */
export { FormKitKitchenSink } from './FormKitKitchenSink'

/**
 * The FormKitMessages component.
 * @public
 */
export { FormKitMessages } from './FormKitMessages'

/**
 * The FormKitProvider component.
 * @public
 */
export {
  FormKitProvider,
  FormKitLazyProvider,
  useConfig,
} from './FormKitProvider'

/**
 * Exports the useFormKitContext composable.
 */
export {
  useFormKitContext,
  useFormKitContextById,
  useFormKitNodeById,
} from './composables/useContext'

/**
 * The FormKitSummary component.
 * @public
 */
export { FormKitSummary, FormKitSummaryMessage } from './FormKitSummary'

/**
 * The FormKitSchema component.
 */
export {
  FormKitSchema,
  Renderable,
  RenderableList,
  RenderableSlot,
  RenderableSlots,
  FormKitComponentLibrary,
  VirtualNode,
} from './FormKitSchema'

/**
 * The default configuration.
 */
export {
  defaultConfig,
  DefaultConfigOptions,
  PluginConfigs,
} from './defaultConfig'

/**
 * The vue specific FormKit core plugin. This is generally required for all
 * vue based FormKit configurations.
 */
export { default as bindings } from './bindings'

/**
 * A vue component for rendering icons from the FormKit icon registry
 */
export { FormKitIcon } from './FormKitIcon'

/**
 * The counter reset for sequential identifiers.
 */
export { resetCount } from './utilities/resetCount'

/**
 * Export the reset count explicitly
 */
export {
  errorHandler,
  setErrors,
  clearErrors,
  submitForm,
  reset,
} from '@formkit/core'

/**
 * Export the changeLocale function explicitly.
 */
export { changeLocale } from '@formkit/i18n'

/**
 * SSR support for cleanup operations relating to SSR.
 */
export { ssrComplete, onSSRComplete } from './composables/onSSRComplete'
