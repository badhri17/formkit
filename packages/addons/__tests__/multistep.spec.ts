import { ConcreteComponent, h, reactive } from 'vue'
import { mount } from '@vue/test-utils'
import {
  FormKit,
  FormKitSchema,
  plugin,
  defaultConfig,
  resetCount,
} from '@formkit/vue'
import { createMultiStepPlugin } from '../src/plugins/multiStep/multiStepPlugin'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

const multiStepSchemaBasic = [
  {
    $formkit: 'multi-step',
    children: [
      {
        $formkit: 'step',
        name: 'stepOne',
        children: [
          {
            $formkit: 'text',
            validation: 'required',
          },
        ],
      },
      {
        $formkit: 'step',
        name: 'stepTwo',
      },
      {
        $formkit: 'step',
        name: 'stepThree',
      },
    ],
  },
]

const multiStepSchemaBasicWithProps = [
  {
    $formkit: 'multi-step',
    tabStyle: 'progress',
    hideProgressLabels: true,
    allowIncomplete: true,
    children: [
      {
        $formkit: 'step',
        name: 'stepOne',
        children: [
          {
            $formkit: 'text',
            validation: 'required',
          },
        ],
      },
      {
        $formkit: 'step',
        name: 'stepTwo',
      },
      {
        $formkit: 'step',
        name: 'stepThree',
      },
    ],
  },
]

describe('multistep', () => {
  beforeEach(() => {
    resetCount()
  })
  afterEach(() => {
    // restore the spy created with spyOn
    vi.restoreAllMocks()
  })

  it('it can mount a multi-step input', () => {
    const wrapper = mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    expect(wrapper.html()).toContain('data-type="multi-step"')
    wrapper.unmount()
  })

  it('it can mount steps inside of a multistep input', () => {
    const wrapper = mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
      slots: {
        default() {
          return h(FormKit as ConcreteComponent, {
            type: 'step',
          })
        },
      },
    })
    expect(wrapper.html()).toContain('data-type="step"')
    wrapper.unmount()
  })

  it('it throws a warning when using a step input outside of a multi-step', () => {
    vi.spyOn(global.console, 'warn').mockImplementation(() => {})

    const wrapper = mount(FormKit, {
      props: {
        type: 'step',
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })
    expect(console.warn).toHaveBeenCalledTimes(1)
    expect(console.warn).toBeCalledWith(
      'Invalid use of <FormKit type="step">. <FormKit type="step"> must be an immediate child of a <FormKit type="multi-step"> component.'
    )
    wrapper.unmount()
  })

  it('it throws a warning when using a non-step input as a direct child of a multi-step', () => {
    vi.spyOn(global.console, 'warn').mockImplementation(() => {})

    const wrapper = mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
      slots: {
        default() {
          return h(FormKit as ConcreteComponent, {
            type: 'text',
          })
        },
      },
    })
    expect(console.warn).toHaveBeenCalledTimes(1)
    expect(console.warn).toBeCalledWith(
      'Invalid FormKit input location. <FormKit type="multi-step"> should only have <FormKit type="step"> inputs as immediate children. Failure to wrap child inputs in <FormKit type="step"> can lead to undesired behaviors.'
    )

    wrapper.unmount()
  })

  it('creates 3 step tabs when it has 3 children of type step', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html().match(/button class=\"formkit-tab\"/g)?.length).toBe(
      3
    )
    wrapper.unmount()
  })

  it('defaults to expected prop arguments', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html()).toContain('data-tab-style="tab"')
    expect(wrapper.html()).toContain('data-hide-labels="false"')
    wrapper.unmount()
  })

  it('accepts new prop argument values', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasicWithProps,
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html()).toContain('data-tab-style="progress"')
    expect(wrapper.html()).toContain('data-hide-labels="true"')
    wrapper.unmount()
  })

  it('Does not allow step advancement when current step is invalid', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [
                createMultiStepPlugin({
                  allowIncomplete: false,
                }),
              ],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 15))
    wrapper.find('.formkit-step-next button').trigger('click')
    await new Promise((r) => setTimeout(r, 15))
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('Allows step advancement when current step is invalid but allowIncomplete is true', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasicWithProps,
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [
                createMultiStepPlugin({
                  allowIncomplete: true,
                }),
              ],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    wrapper.find('.formkit-step-next button').trigger('click')
    await new Promise((r) => setTimeout(r, 15))
    // 2nd tab is active (without labels due to props)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('preserves the order of steps even when a step is conditionally rendered', async () => {
    const data = reactive({
      showStepTwo: true,
    })
    const wrapper = mount(FormKitSchema, {
      props: {
        data,
        schema: [
          {
            $formkit: 'multi-step',
            children: [
              {
                $formkit: 'step',
                name: 'stepAlpha',
                key: 'stepOne',
              },
              {
                $formkit: 'step',
                if: '$showStepTwo',
                name: 'stepBravo',
                key: 'stepTwo',
              },
              {
                $formkit: 'step',
                name: 'stepCharlie',
                key: 'stepThree',
              },
            ],
          },
        ],
      },
      attachTo: document.body,
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    const stepNameRegex = /Step (.*)?</gm
    await new Promise((r) => setTimeout(r, 15))
    const stepMatches = wrapper.html().match(stepNameRegex)
    expect(stepMatches).toEqual(['Step Alpha<', 'Step Bravo<', 'Step Charlie<'])
    data.showStepTwo = false
    await new Promise((r) => setTimeout(r, 15))
    const stepMatchesAfter = wrapper.html().match(stepNameRegex)
    expect(stepMatchesAfter).toEqual(['Step Alpha<', 'Step Charlie<'])
    data.showStepTwo = true
    await new Promise((r) => setTimeout(r, 20))
    const stepMatchesAfter2 = wrapper.html().match(stepNameRegex)
    expect(stepMatchesAfter2).toEqual([
      'Step Alpha<',
      'Step Bravo<',
      'Step Charlie<',
    ])
    wrapper.unmount()
  })
})
