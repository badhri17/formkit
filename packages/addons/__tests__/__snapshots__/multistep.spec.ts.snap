// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multistep > Allows step advancement when current step is invalid but allowIncomplete is true 1`] = `
"<div class=\\"formkit-outer\\" id=\\"input_0\\" data-family=\\"multi-step\\" data-type=\\"multi-step\\">
  <div class=\\"formkit-wrapper\\" data-tab-style=\\"progress\\" data-hide-labels=\\"true\\">
    <div class=\\"formkit-tabs\\" role=\\"tablist\\"><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"false\\" data-visited=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_0\\" aria-selected=\\"false\\" aria-controls=\\"input_1\\" tabindex=\\"-1\\">
        <!----><span class=\\"formkit-badge\\" role=\\"presentation\\">1</span>
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"true\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_1\\" aria-selected=\\"true\\" aria-controls=\\"input_4\\" tabindex=\\"0\\" data-visited=\\"true\\">
        <!---->
        <!----><span class=\\"formkit-badge\\" role=\\"presentation\\"><!----></span>
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_2\\" aria-selected=\\"false\\" aria-controls=\\"input_6\\" tabindex=\\"-1\\">
        <!---->
        <!---->
        <!---->
      </button></div>
    <div class=\\"formkit-steps\\">
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_1\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_0\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <div class=\\"formkit-outer\\" data-family=\\"text\\" data-type=\\"text\\" data-empty=\\"true\\" data-invalid=\\"true\\" data-submitted=\\"true\\">
            <div class=\\"formkit-wrapper\\">
              <!---->
              <div class=\\"formkit-inner\\">
                <!---->
                <!----><input class=\\"formkit-input\\" type=\\"text\\" name=\\"text_2\\" id=\\"input_2\\" aria-required=\\"true\\" aria-describedby=\\"input_2-rule_required\\">
                <!---->
                <!---->
              </div>
            </div>
            <!---->
            <ul class=\\"formkit-messages\\">
              <li class=\\"formkit-message\\" id=\\"input_2-rule_required\\" data-message-type=\\"validation\\">Text_2 is required.</li>
            </ul>
          </div>
        </div>
        <div class=\\"formkit-step-actions\\">
          <!---->
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_3\\" id=\\"input_3\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_4\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_1\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_4\\" id=\\"input_5\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_6\\" id=\\"input_8\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_6\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_2\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_5\\" id=\\"input_7\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`multistep > Does not allow step advancement when current step is invalid 1`] = `
"<div class=\\"formkit-outer\\" id=\\"input_0\\" data-family=\\"multi-step\\" data-type=\\"multi-step\\">
  <div class=\\"formkit-wrapper\\" data-tab-style=\\"tab\\" data-hide-labels=\\"false\\">
    <div class=\\"formkit-tabs\\" role=\\"tablist\\"><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"true\\" data-valid=\\"false\\" data-visited=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_0\\" aria-selected=\\"true\\" aria-controls=\\"input_1\\" tabindex=\\"0\\"><span class=\\"formkit-tab-label\\">Step One</span><span class=\\"formkit-badge\\" role=\\"presentation\\">1</span>
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_1\\" aria-selected=\\"false\\" aria-controls=\\"input_4\\" tabindex=\\"-1\\"><span class=\\"formkit-tab-label\\">Step Two</span>
        <!---->
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_0_tab_2\\" aria-selected=\\"false\\" aria-controls=\\"input_6\\" tabindex=\\"-1\\"><span class=\\"formkit-tab-label\\">Step Three</span>
        <!---->
        <!---->
      </button></div>
    <div class=\\"formkit-steps\\">
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_1\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_0\\">
        <div class=\\"formkit-step-inner\\">
          <div class=\\"formkit-outer\\" data-family=\\"text\\" data-type=\\"text\\" data-empty=\\"true\\" data-invalid=\\"true\\" data-submitted=\\"true\\">
            <div class=\\"formkit-wrapper\\">
              <!---->
              <div class=\\"formkit-inner\\">
                <!---->
                <!----><input class=\\"formkit-input\\" type=\\"text\\" name=\\"text_2\\" id=\\"input_2\\" aria-required=\\"true\\" aria-describedby=\\"input_2-rule_required\\">
                <!---->
                <!---->
              </div>
            </div>
            <!---->
            <ul class=\\"formkit-messages\\">
              <li class=\\"formkit-message\\" id=\\"input_2-rule_required\\" data-message-type=\\"validation\\">Text_2 is required.</li>
            </ul>
          </div>
        </div>
        <div class=\\"formkit-step-actions\\">
          <!---->
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_3\\" id=\\"input_3\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_4\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_1\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_4\\" id=\\"input_5\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_6\\" id=\\"input_8\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_6\\" role=\\"tabpanel\\" aria-labelledby=\\"input_0_tab_2\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\" data-empty=\\"true\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_5\\" id=\\"input_7\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>"
`;
